app/src/main/java/com/vci/vectorcamapp/core/data/upload/image/ImageUploadWorker.kt

```diff
@@ -2,85 +2,508 @@ package com.vci.vectorcamapp.core.data.upload.image
+                                    Log.e(
+                                        "ImageUploadWorker",
+                                        "Non-retryable error for specimen ${specimen.id} on attempt $attempt. This is a permanent failure for this image.",
+                                    )
+                                    permanentErrorOccurred = true
+                                }
+                            }
+                        }
+                    } catch (e: Exception) {
+                        Log.e(
+                            "ImageUploadWorker",
+                            "Exception on attempt $attempt for specimen ${specimen.id}. This is a permanent failure for this image.",
+                            e
+                        )
+                        permanentErrorOccurred = true
+                    }
+                }
+
+                if (!imageUploadSuccess) {
+                    Log.e(
+                        "ImageUploadWorker",
+                        "All attempts failed or a permanent error occurred for specimen ${specimen.id}. Moving to next image."
+                    )
+                    val updatedSpecimen = specimen.copy(imageUploadStatus = UploadStatus.FAILED)
+                    specimenRepository.updateSpecimen(updatedSpecimen, sessionId)
+                }
+            } finally {
+                file.delete()
             }
-        } catch (e: Exception) {
-            Log.d("UploadWorker", "Error during upload: ${e.message}")
-            return Result.retry()
         }
 
-        return Result.success()
+        showFinalStatusNotification(sessionId.toString(), successfulUploads, specimensToUpload.size)
+        return WorkerResult.success()
     }
 
-    private fun createNotificationChannel() {
-        val channel = NotificationChannel(
-            CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_LOW
+    private suspend fun performUpload(
+        file: File,
+        contentType: String?,
+        uploadSpecimenId: String
+    ): DomainResult<String, NetworkError> {
+        val md5 = calculateMD5(file)
+        val uniqueFingerprint = "$uploadSpecimenId-$md5"
+        val tusPath = "specimens/$uploadSpecimenId/images/tus"
+
+        tusClient.uploadCreationURL = URL(constructUrl(tusPath))
+        val upload = createTusUpload(file, uniqueFingerprint, contentType, md5)
+
+        Log.d("ImageUploadWorker", "Start/resume ${file.name} (fp=$uniqueFingerprint,md5=$md5)")
+
+        val uploaderResult = try {
+            DomainResult.Success(tusClient.resumeOrCreateUpload(upload))
+        } catch (e: TusProtocolException) {
+            if (e.causingConnection?.responseCode == HttpURLConnection.HTTP_CONFLICT) {
+                Log.w(
+                    "ImageUploadWorker",
+                    "resumeOrCreateUpload failed with 409 Conflict. Verifying by MD5.",
+                    e
+                )
+                return when (val existingUrlResult =
```

Comment:
I feel like you can assume that if the backend sends a conflict, then the image was initialized. You dont need to verify with MD5 check